src/jrd/dpm.epp

```diff
@@ -2579,19 +2579,24 @@ static void delete_tail(thread_db* tdbb, rhdf* header, const USHORT page_space,
 
 	if (!(header->rhdf_flags & rhd_blob))
 	{
+		USHORT flags = header->rhdf_flags;
 		ULONG page_number = header->rhdf_f_page;
-		while (true)
+
+		while (flags & rhd_incomplete)
 		{
 			window.win_page = page_number;
 			data_page* dpage = (data_page*) CCH_FETCH(tdbb, &window, LCK_read, pag_data);
-			header = (rhdf*) ((UCHAR *) dpage + dpage->dpg_rpt[0].dpg_offset);
-			const USHORT flags = header->rhdf_flags;
-			page_number = header->rhdf_f_page;
 
-			CCH_RELEASE_TAIL(tdbb, &window);
-			PAG_release_page(tdbb, window.win_page, ZERO_PAGE_NUMBER);
+			if (dpage->dpg_header.pag_flags & dpg_orphan)
```

Comment:
Could you explain this condition ?

Reply:
The code used to release data pages allocated for fragments of big records when relation is dropped.
Fragments are allocated on own separate data pages and is not marked as orphan.
How this new code is going to release fragment's pages ?

Reply:
Sorry, fragment's page is marked as orphan, I missed it.
But it could be better (safer) to check also `dpg_full` flag and number of records on page, same as in `analyze_fragments()` at dba.epp

Reply:
> Fragments are allocated on own separate data pages and is not marked as orphan.
> How this new code is going to release fragment's pages ?

If a data page is not an orphan then it's included in a pointer page. Such pages will be released in DPM_delete_relation_pages function (note that it checks rhd_large flag for secondary data pages as well). It's not the task of delete_tail function.

Reply:
> But it could be better (safer) to check also `dpg_full` flag and number of records on page, same as in `analyze_fragments()` at dba.epp

I agree. But I just looked at `analyze_fragments()` and the condition there is a bit confusing for me. If a data page has `dpg_full` and `dpg_count = 1` but doesn't have `dpg_orphan`, the condition is still `true`, right?

Reply:
Yes, `analyze_fragments()` is not as accurate (or paranoid) as it might be ;) 
But you get the idea - lets check for all 3 conditions to be absolutely sure.